File: src/206ReverseLinkedList/206ReverseLinkedList.cpp
#include <iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *reverseList(ListNode *head) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode *r = &dummy;
    ListNode *p = head;
    while (p != nullptr) {
      ListNode *q = p->next;
      p->next = r;
      r = p;
      p = q;
    }

    if (dummy.next != nullptr) {
      dummy.next->next = nullptr;
      return r;
    } else {
      return nullptr;
    }
  }
};
int main() {}

File: src/206ReverseLinkedList/206ReverseLinkedList.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut p = None;
        let mut q = head;
        while let Some(qn) = q.as_mut() {
            let r = qn.next.take();
            qn.next = p;
            p = q;
            q = r;
        }
        p
    }
}
